src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetProjectManagerService.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft;
+using Microsoft.ServiceHub.Framework;
+using Microsoft.ServiceHub.Framework.Services;
+using NuGet.Packaging;
+using NuGet.ProjectManagement;
+using NuGet.VisualStudio;
+using NuGet.VisualStudio.Internal.Contracts;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    public class NuGetProjectManagerService : INuGetProjectManagerService
+    {
+        private bool _disposedValue;
+        private readonly ServiceActivationOptions _options;
+        private readonly IServiceBroker _serviceBroker;
+        private readonly AuthorizationServiceClient _authorizationServiceClient;
+
+        public NuGetProjectManagerService(ServiceActivationOptions options, IServiceBroker sb, AuthorizationServiceClient ac, CancellationToken ct)
+        {
+            _options = options;
+            _serviceBroker = sb;
+            _authorizationServiceClient = ac;
+        }
+
+        public async ValueTask<IReadOnlyCollection<PackageReference>> GetInstalledPackagesAsync(IReadOnlyCollection<string> projectGuids, CancellationToken ct)
+        {
+            var solutionManager = await ServiceLocator.GetInstanceAsync<IVsSolutionManager>();
+            Assumes.NotNull(solutionManager);
+
+            var projects = (await solutionManager.GetNuGetProjectsAsync()).Where(p => projectGuids.Contains(p.GetMetadata<string>(NuGetProjectMetadataKeys.ProjectId))).ToArray();
+
+            // Read package references from all projects.
+            var tasks = projects.Select(project => project.GetInstalledPackagesAsync(ct));
+            var packageReferences = await Task.WhenAll(tasks);
+
+            // Group all package references for an id/version into a single item.
+            return packageReferences.SelectMany(e => e).ToArray();
```

Comment:
I didn't see a `PackageReference.Equals`, so just checking... does this always work? I see the Version and ID will be grouped, but does this expect 2 groups for the same ID/Version if say, `IsDevelopmentDependency` is different on one?

Reply:
Ah the comment was from something else and does not apply here, we are not grouping by id/version here, this is just flattening a collection of lists into one single list.